### ISSUE TITLE
state boundaries have gaps

### ISSUE THREAD

**Issue by cldellow** [systemed/tilemaker#545]
I'm trying to render a North American map with the latest code in master, and have noticed that many state borders have gaps. For example, Kentucky ([relation #161655](https://www.openstreetmap.org/relation/161655#map=7/37.834/-85.768)) renders like so:
![Selection_594](https://github.com/systemed/tilemaker/assets/193185/baff5d21-9d6f-4358-997a-63a4b367db6d)

Note the small gap in its lower left and big gap in its lower right.

I ran `ogrinfo -al` on the resulting mbtiles file as well as on the OSM PBF. You can [see that here](https://gist.github.com/cldellow/c691485a9aefd6398c10998d8edfaddd) -- the OSM PBF has a _lot_ more data points in it than the thing that got rendered.

Reading through past issues in this repo, I suspect this may be related to multipolygon support. But I'm a GIS newbie, so now I'm stuck. :) Do you have any hints on how I could troubleshoot further/fix this?

PS - thanks for this amazing tool!

**Comment by cldellow** [comment]
I saw that Tilemaker has [pretty good support for shapefiles](https://github.com/systemed/tilemaker/blob/master/docs/CONFIGURATION.md#shapefiles), so I tried a workaround: export boundaries to a shapefile, then use those. This seems to avoid the issue. I don't understand why. Perhaps GDAL silently fixes something? Perhaps shapefiles don't have full fidelity to the original PBF? It's a glorious mystery to me! Maybe one day I'll understand enough to offer an actual PR, but for now, here's the workaround in case it's useful to anyone:

1.  Create the shapefile
```bash
# pre-filter the PBF file to only have boundary relations -- otherwise ogr2ogr takes 10x as long
osmium tags-filter north-america-latest.osm.pbf -o north-america-latest-filtered.osm.pbf  "r/admin_level"

# export country/state borders - coerce the admin_level to a number before comparing it
ogr2ogr -dialect SQLite -sql "select * from multipolygons where boundary = 'administrative' and 1 * admin_level <= 4" boundaries/na-borders.shp north-america-latest-filtered.osm.pbf
```

2. Tweak your `config.json` to read this shapefile and write it to the `boundary` layer. Use `"source_columns": true` so that the original OSM tags are available to the Lua processing script:

```json
    "na_borders":             { "minzoom": 0,  "maxzoom": 14, "source": "boundaries/na-borders.shp", "simplify_below": 13, "simplify_level": 0.0001, "simplify_ratio": 2, "write_to": "boundary", "index": true, "source_columns": true },
```

3. Adjust the `attribute_function` in the Lua processing script to handle this new shapefile. Shapefiles have a maximum length for their attributes, so `admin_level` gets truncated to `admin_leve`. Fix that up in order to maintain the name that OpenMapTiles schema expects:

```lua
function attribute_function(attr,layer)
  if layer == 'na_borders' then
    return {
      osm_id=attr['osm_id'],
      admin_level=tonumber(attr['admin_leve']),
    }
  end

  -- processing for other shapefiles
  ...
end
```

This results in nice looking borders: 
![Selection_601](https://github.com/systemed/tilemaker/assets/193185/e9a3460a-c212-4844-aaac-8070c4250053)

**Comment by systemed** [comment]
> Reading through past issues in this repo, I suspect this may be related to multipolygon support. But I'm a GIS newbie, so now I'm stuck. :) Do you have any hints on how I could troubleshoot further/fix this?

Out of interest, are you using the default openmaptiles-compatible schema (JSON/Lua), or have you rolled your own?

**Comment by cldellow** [comment]
This was using the default openmaptiles config/process files that are checked into the master branch.

Your question led me to start writing out more details on what I had attempted, which then led to me doing a rubber duck debug session with myself. Rubber duck session below in case it helps anyone else who's starting to poke around Tilemaker.

tl;dr: I now think the issue is this line in the process.lua script: https://github.com/systemed/tilemaker/blame/439c6b10276a8427466a1e5f3e54df4c79361d1e/resources/process-openmaptiles.lua#L270

Deleting that line seems to resolve my issue. This sort of makes sense to me -- we just did a bunch of work to compute the admin_level based on the relations that the way is a member of, and now we're discarding it in favour of an admin_level set directly on the way.

I don't have enough context to know if removing that line would break something else, though. It looks like this was introduced fairly recently as part of the boundary support changes in https://github.com/systemed/tilemaker/commit/528aa25deca1181d66af01ea5b954ffc0abd34a8 

__My notes as I was writing this response__:

The OSM data was initially from a small extract of Kentucky and surrounding area, generated from BBBike.

I then wondered if I had somehow chosen a bounding box that clipped out needed parts of the borders, so I tried a few other things:

I found a few ways that are used by the Kentucky relation, some that were included in the final map output (good) and some that weren't (bad):

- [785710532](https://www.openstreetmap.org/way/785710532) - rendered -- this has an explicit `admin_level` field
- [46148135](https://www.openstreetmap.org/way/46148135) - not rendered
- [946699500](https://www.openstreetmap.org/way/946699500) - not rendered
- [224301862](https://www.openstreetmap.org/way/224301862) - not rendered

All of these ways were present in my extract when I looked for them with `osmium getid -f opl extract.osm.pbf w785710532 ...`

Given that, I didn't think the issue was with my extract. Still, I tried again with a Geofabrik export of the entire US. The issue persisted, though.

Ah, I should go put some debug code in the process.lua for way 785710532 and 46148135 and see how they differ.

...right, 46148135 gets output with an `admin_level` of 11, whereas 785710532 has an `admin_level` of 4.

This seems like a recent change in the process.lua code, as part of the rejigging of boundary polygons in https://github.com/systemed/tilemaker/commit/528aa25deca1181d66af01ea5b954ffc0abd34a8

**Comment by systemed** [comment]
That's spot on and thanks for debugging that. It should probably be

    admin_level = math.min(admin_level, tonumber(way:Find("admin_level")) or 11)

**Comment by systemed** [comment]
#549 addresses this.
